src/utils/errorHandling.ts

```diff
@@ -0,0 +1,49 @@
+import logger from '@/logger';
+import type { Success, Failure, ApiFailure } from '@/shared.types';
+
+function createSuccess<T>(data?: T): Success<T> {
+  return { success: true, data };
+}
+
+async function getResponseError(response: Response): Promise<string> {
+  const body = await response.json();
+  return body.error ? body.error : 'Unknown error';
+}
+
+// Adding the option to return a ApiFailure Result type rather than
+// throwing an error because different functions might want to
+// handle response status codes differently. E.g., 404 might be an
+// error sometimes, but for preferences, it just means that preference
+// key hasn't been set yet.
```

Comment:
Could this be handled with exceptions in the business logic? I'm not sure this kind of decision needs to be bubbled up to the UI. Is there an example?

Reply:
I think the underlying issue is that `fetchTicket` is missing a return type. It returns a Result when there is an error, but a bare object when there is success. If we set the return type to `Promise<Result<Ticket | null>>`, I think the type system would force solutions to a lot of problems, which is our goal with this refactoring. For example, it would be clear that this function must be responsible for parsing the HTTP return code and deciding what/how to log and what to return to the user. Also, the `useEffect` (hopefully) won't need to try/catch.